### ISSUE TITLE
RPC: cannot select from glare_catalog

### ISSUE THREAD

**Issue by greyscaled** [GlareDB/glaredb#1657]
I'm unsure if this is part of an existing issue, but wanted to document it explicitly for tracking purposes.

**Comment by scsmithr** [comment]
Fix for this is adding:

```
        let provider = if let Some(_) = self.ctx.exec_client() {
            Arc::new(LocalTableHint(provider))
        } else {
            provider
        };
```

before the return here: https://github.com/glaredb/glaredb/blob/ab4e712e115183609386242df0b7d23c89950dc1/crates/sqlexec/src/planner/context_builder.rs#L154

I think we'll want to look at alternatives though. Currently we default to assuming everything runs remote except when wrapped in a `LocalTableHint`. What I think we should do is add a `LocationAwareTableProvider` that explicitly says if it should run locally or remote (or doesn't matter). By ensuring everything is a `LocationAwareTableProvider`, we would be able to more easily optimize these queries by checking the locations for providers. E.g. if all table providers are "local", then just run everything local and don't attempt to send to remote node.

If we go in this direction, I would want us to remove returning `Arc<dyn TableProvider>` from everywhere that we return a table provider, and instead return `LocationAwareTableProvider`. This would ensure we have location info for all tables we're working with.

cc @universalmind303

**Comment by vrongmeal** [comment]
I think a bigger issue is deciding whether or not to create the table provider locally or not. Right now we `dispatch_access` everything on remote. We would never get a table provider with a `LocalTableHint`. We can do a pretty good job with the `PartialContextProvider` itself without making many changes to the code. Later on we can optimize query based on location. First, I think we should figure this out.

Basically, I am ok with sending a completely local query to remote and getting the exact same result but we need to refine our `dispatch_access`. `LocalTableHint` can be wrapped by `PartialContextProvider` (and hopefully not required anywhere else) and we'll have full control over what scans on local and what scans on remote.

**Comment by vrongmeal** [comment]
Same goes for functions. We might have to modify the `AsyncContextProvider` (or `TableFunc`) a bit to have more control over where to execute the function but this will me much simpler.

@universalmind303 ^

**Comment by scsmithr** [comment]
I edited the issue+title to be more about how we want to plan local vs remote. The original issues is "solved", but there's some good discussion here that I didn't want to lose when it comes to being more versatile with planning.

**Comment by universalmind303** [comment]
@scsmithr @vrongmeal is there anything actionable left on this issue? I feel like most of this was resolved in 0.5.0.

**Comment by greyscaled** [comment]
Our planning is robust enough for an early iteration. Going to close.

We need to work on physical plans that don't get serialized (#1698) and explains (#1645)
